src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,648 @@
+    for (auto const& [premises, conclusion] : *forbidden_rules) {
+        if (CompareLiteralSets(premises, subset) && ContainsLiteral(conclusion, l)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+void HGenerate(std::vector<Literal>& notsatisfied, Literal const& l, std::vector<SingleRule>& rules,
+               graph_t const& graph, graph_t& pattern, Embeddings const& embeddings,
+               std::vector<Rules>::iterator const& forbidden_rules, std::size_t const sigma) {
+    std::vector<Literal> conclusion = {l};
+    for (std::size_t k = 1; k <= 3; ++k) {
+        std::vector<std::vector<Literal>> subsets = GetSubsets<Literal>(notsatisfied, k);
+        for (auto& subset : subsets) {
+            if (IsForbidden(forbidden_rules, subset, l)) {
+                continue;
+            }
+
+            Gfd gfd = {pattern, subset, conclusion};
+            if (Validate(graph, gfd, embeddings, sigma)) {
+                rules.push_back({subset, l});
+                for (auto const& lit : subset) {
+                    auto last = std::remove(notsatisfied.begin(), notsatisfied.end(), lit);
+                    notsatisfied.erase(last, notsatisfied.end());
+                }
+            }
+        }
+    }
+}
+
+std::vector<SingleRule> GenerateRules(std::vector<Literal> const& literals, graph_t const& graph,
+                                      graph_t& pattern, Embeddings const& embeddings,
+                                      std::vector<Rules>::iterator const& forbidden_rules,
+                                      std::size_t const sigma) {
+    std::vector<SingleRule> rules = {};
+    for (std::size_t j = 0; j < literals.size(); ++j) {
+        std::vector<Literal> current = literals;
+        Literal l = current.at(j);
+        current.erase(std::next(current.begin(), j));
+
+        if (IsForbidden(forbidden_rules, {}, l)) {
+            continue;
+        }
+
+        std::vector<Literal> conclusion = {l};
+        std::vector<Literal> premises = {};
+        Gfd gfd = {pattern, premises, conclusion};
+        if (Validate(graph, gfd, embeddings, sigma)) {
+            rules.push_back({{}, l});
+            continue;
+        }
+        std::vector<Literal> notsatisfied = current;
+        HGenerate(notsatisfied, l, rules, graph, pattern, embeddings, forbidden_rules, sigma);
+    }
+    return rules;
+}
+
+void AddCompacted(std::vector<SingleRule>& rules, std::vector<Gfd>& result, graph_t& pattern,
+                  std::vector<Rules>::iterator& forbidden_rules) {
+    if (!rules.empty()) {
+        std::sort(rules.begin(), rules.end());
+        std::vector<Literal> prev = rules.front().first;
+        std::vector<Literal> conclusion = {};
+        for (auto const& [premises, l] : rules) {
+            if (prev != premises) {
+                Gfd gfd = {pattern, prev, conclusion};
+                result.push_back(std::move(gfd));
+                forbidden_rules->push_back({prev, conclusion});
```

Comment:
To avoid creating a temporary object use `emplace_back` with constructor parameter to create an object right in a container.

```suggestion
                forbidden_rules->emplace_back(prev, conclusion);
```
Applies everywhere else